sqlx-core/src/postgres/types/decimal.rs

```diff
@@ -0,0 +1,443 @@
+        PgTypeInfo::NUMERIC
+    }
+}
+
+impl Type<Postgres> for [Decimal] {
+    fn type_info() -> PgTypeInfo {
+        PgTypeInfo::NUMERIC_ARRAY
+    }
+}
+
+impl Type<Postgres> for Vec<Decimal> {
+    fn type_info() -> PgTypeInfo {
+        <[Decimal] as Type<Postgres>>::type_info()
+    }
+}
+
+impl TryFrom<PgNumeric> for Decimal {
+    type Error = BoxDynError;
+
+    fn try_from(numeric: PgNumeric) -> Result<Self, BoxDynError> {
+        let (digits, sign, weight) = match numeric {
+            PgNumeric::Number {
+                digits,
+                sign,
+                weight,
+                ..
+            } => (digits, sign, weight),
+
+            PgNumeric::NotANumber => {
+                return Err("Decimal does not support NaN values".into());
+            }
+        };
+
+        if digits.is_empty() {
+            // Postgres returns an empty digit array for 0 but BigInt expects at least one zero
+            return Ok(0u64.into());
+        }
+
+        let sign = match sign {
+            PgNumericSign::Positive => Sign::Plus,
+            PgNumericSign::Negative => Sign::Minus,
+        };
+
+        // weight is 0 if the decimal point falls after the first base-10000 digit
+        let scale = (digits.len() as i64 - weight as i64 - 1) * 4;
+
+        // no optimized algorithm for base-10 so use base-100 for faster processing
+        let mut cents = Vec::with_capacity(digits.len() * 2);
+        for digit in &digits {
+            cents.push((digit / 100) as u8);
+            cents.push((digit % 100) as u8);
+        }
+
+        let bigint = BigInt::from_radix_be(sign, &cents, 100)
+            .ok_or("PgNumeric contained an out-of-range digit")?;
+
+        match bigint.to_i128() {
+            Some(num) => Ok(Decimal::from_i128_with_scale(num, scale as u32)),
+            None => Err("Decimal's integer part out of range.".into()),
+        }
+    }
+}
+
+impl TryFrom<&'_ Decimal> for PgNumeric {
+    type Error = BoxDynError;
+
+    fn try_from(decimal: &Decimal) -> Result<Self, BoxDynError> {
+        if decimal.is_zero() {
+            return Ok(PgNumeric::Number {
+                sign: PgNumericSign::Positive,
+                scale: 0,
+                weight: 0,
+                digits: vec![],
+            });
+        }
+
+        let scale = decimal.scale() as u16;
+        let groups_diff = scale & 0x3;
+
+        // A serialized version of the decimal number. The resulting byte array
+        // will have the following representation:
+        //
+        // Bytes 1-4: flags
+        // Bytes 5-8: lo portion of m
+        // Bytes 9-12: mid portion of m
+        // Bytes 13-16: high portion of m
+        let s = decimal.serialize();
+
+        let mut mantissa = [
+            u32::from_le_bytes(s[4..8].try_into().unwrap()),
+            u32::from_le_bytes(s[8..12].try_into().unwrap()),
+            u32::from_le_bytes(s[12..16].try_into().unwrap()),
+            0u32,
+        ];
```

Comment:
As I mentioned in Discord, I think this all gets a lot simpler if we use `u128` for the math. You don't even have to do anything fancy to convert the serialization, just use `u128::from_le_bytes()` and then mask off the top 32 bits:

```rust
let mut mantissa = u128::from_le_bytes(decimal.serialize());
// the top 32 bits are for flags and scale so we chop those off
mantissa &= (1u128 << 96) - 1;

// if our scale is not a multiple of 4 we need to go to the next multiple
if groups_diff > 0 {
    mantissa = mantissa * 10u32.pow(4 - groups_diff as u32);
}

while mantissa != 0 {
    digits.push((mantissa % 10_000) as u16);
    mantisa /= 10_000;
}
```

Reply:
It _should_ perform the same, although in my black-box testing it looked like LLVM failed to combine the division and modulo optimizations into one library call.

Reply:
With using u128 I was able to get completely wrong results. Should we try it out still, or go with this approach with u96?

Reply:
Wow, completely wrong?

Reply:
I can try it again later today and show you. I probably missed something stupid in my try, but yeah got all tests red.

Now when I added some more tests, I found out more problems with the implementation. This time to the other direction, from database to `Decimal`. The problem there is the scale calculation... `Decimal` uses `u32` and `BigDecimal` uses `i32`. I'm able to overflow the scale with `Decimal::new(10000, 0)`, which means the whole function needs to be thought out again.

Maybe taking the approach from the crate itself: https://github.com/paupino/rust-decimal/blob/master/src/postgres.rs#L34

That thing definitely needs more comments, and I really hate how much I suck with this. I might need to read more to really understand what's going on here. Would be better to get the internals of decimals finally...

Reply:
https://github.com/launchbadge/sqlx/pull/456/commits/1f7b71b4e2660e4c7c594d9c8229766e65991895 This tries to address the overflowing scale in the reading part.

Reply:
Ok, so with the u128 version over here, I was able to get almost all tests working, except when the number is `1`. This means the mantissa overflows to `4294967296` and the test fails. Trying to debug this before the 4th of July BBQ :)

Reply:
Ha. Finally kind of got how to do this. Went with the u128 math ops, and found a problem from your suggestion I understood and fixed! Yay!

So you can see the commit here: https://github.com/launchbadge/sqlx/pull/456/commits/6360558a27e30ad70c0151562b485fdb22dd3a7c

What the example above thinks is the flags are in the end. But actually they're the first four bytes, then we have a u96 as little-endian after that. Maybe I could do my conversion in a more beautiful way, but what I did here is I move the parts of the number to the front, zero out the flags and convert the bytes to a u128. No need to chop off the flags and all tests green.

Please read the changes and you might find some parts I can do even prettier.